signingscript/src/signingscript/sign.py

```diff
@@ -1633,3 +1625,71 @@ async def apple_notarize_geckodriver(context, path, *args, **kwargs):
     utils.mkdir(notarization_workdir)
 
     return await _notarize_geckodriver(context, path, notarization_workdir)
+
+
+@time_async_function
+async def apple_app_hardened_sign(context, path, *args, **kwargs):
+    """
+    Sign an app using rcodesign.
+    """
+    # Setup workdir
+    signing_dir = os.path.join(context.config["work_dir"], "extracted")
+    shutil.rmtree(signing_dir, ignore_errors=True)
+    utils.mkdir(signing_dir)
+
+    if not path.endswith((".dmg", ".tar.gz")):
+        raise SigningScriptError("File format not supported.")
+
+    extension = os.path.splitext(path)[-1]
+    # For now, we convert to tar, then continue work as tar
+    if extension == ".dmg":
+        await utils.extract_dmg(context, path, signing_dir)
+    else:
+        await _extract_tarfile(context, path, extension, signing_dir)
+
+    # Get configs from task payload
+    payload = context.task.get("payload", {})
+    hardened_sign_config = payload.get("hardened-sign-config")
+    assert hardened_sign_config and isinstance(hardened_sign_config, list)
+    provisioning_profile_config = payload.get("provisioning-profile-config")
+
+    signed = False
+    for file in os.scandir(signing_dir):
+        if file.is_dir() and file.name.endswith(".app"):
+            # Developer ID Application certificate
+            creds = context.apple_app_signing_creds_path
+        elif file.is_file() and file.name.endswith(".pkg"):
+            # Use installer credentials
+            creds = context.apple_installer_signing_creds_path
+        else:
+            # If not pkg AND not a directory (.app) - then skip file
```

Comment:
todo: log any files that are skipped to make debugging easier.

Reply:
Done